File: Apps/Codaxy.Dextop.Showcase/Demos/Live/LiveGroupingGridWindow.cs
using System;
using System.Collections.Generic;
using System.Threading;
using Codaxy.Dextop.Data;
using System.Diagnostics;


namespace Codaxy.Dextop.Showcase.Demos.Live
{
    [Demo("LiveGroupingGridWindow",
        Title = "Live Grid Grouping",
        Description = "Grid with live data and grouping.",
        Path = "~/Demos/Live"
    )]
    [LevelMedium]
    [TopicDextopLive]
    [CategoryCodeSnippet]
    public class LiveGroupingGridWindow : DextopWindow
    {
        DextopObservableStore<string, LiveGridModel> store;
        Timer timer;
        Random r;

        public LiveGroupingGridWindow()
        {
            store = new DextopObservableStore<string, LiveGridModel>(a => a.Ticker);
            r = new Random();
#if DEBUG
            timer = new Timer(OnTimer, null, 100, 300);            
#else
            timer = new Timer(OnTimer, null, 100, 2000);            
#endif
        }

        void OnTimer(object state)
        {
            try
            {                
                var changes = new List<LiveGridModel>();
                var n = r.Next(100) + 1;
                for (var i = 0; i < n; i++)
                {
                    changes.Add(new LiveGridModel
                    {
                        Ticker = "Ticker " + r.Next(200),
                        Group = "G" + r.Next(10),
                        Price = Math.Round(r.NextDouble() * 1000, 2)
                    });
                }
                store.SetMany(changes);
            }
            catch { }
        }

        public override void Dispose()
        {
            if (timer != null)
            {
                timer.Dispose();
                timer = null;
            }
            base.Dispose();
        }

        public override void InitRemotable(DextopRemote remote, DextopConfig config)
        {
            base.InitRemotable(remote, config);                      
            Remote.AddLiveStore("model", store);
        }

        [DextopModel]
        [DextopGrid]
        class LiveGridModel
        {
            public String Group { get; set; }

            [DextopModelId]
            [DextopGridColumn(flex=1)]
            public String Ticker { get; set; }
            
            [DextopGridColumn(width = 100)]            
            public double Price { get; set; }
        }
    }
}

File: Apps/Codaxy.Dextop.Showcase/Demos/Live/LiveGroupingGridWindow.js
Ext.ns('Showcase');

Ext.define('Showcase.demos.LiveGroupingGridWindow', {
    extend: 'Dextop.Window',
    width: 500,
    height: 300,

    title: 'Live Grouping Grid',
    requires: 'Ext.grid.GridPanel',
    maximizable: true,

    initComponent: function () {
        var store = this.remote.createLiveStore('model', {
            autoLoad: true,
            autoSort: true,            
            sorters: [{
            	property: 'Ticker',
                direction: 'DESC'
            }],
            groupField: 'Group'
        });

        var columns = this.remote.createGridColumns('model');

        var grid = Ext.create('Ext.grid.GridPanel', {
            store: store,
            columns: columns,
            features: [{ ftype: 'grouping' }],
            viewConfig: {
                loadMask: false
            }
        });

        Ext.apply(this, {
            layout: 'fit',
            items: grid
        });

        this.callParent(arguments);

    }
});
